src/prefect/agent/kubernetes/agent.py

```diff
@@ -191,6 +192,72 @@ def manage_jobs(self) -> None:
+                        pods = self.core_client.list_namespaced_pod(
+                            namespace=self.namespace,
+                            label_selector="prefect.io/identifier={}".format(
+                                job.metadata.labels.get("prefect.io/identifier")
+                            ),
+                        )
+
+                        failed_pods = []
+                        for pod in pods.items:
+                            if pod.status.phase != "Failed":
+                                continue
+
+                            # Format pod failure error message
+                            failed_pods.append(pod.metadata.name)
+                            pod_failure_message = f"Pod {pod.metadata.name} failed.\n"
+                            for status in pod.status.container_statuses:
+                                state = (
+                                    "running"
+                                    if status.state.running
+                                    else "waiting"
+                                    if status.state.waiting
+                                    else "terminated"
+                                    if status.state.terminated
+                                    else "Not Found"
+                                )
+                                msg = f"\tContainer '{status.name}' state: {state}\n"
+
+                                if status.state.terminated:
+                                    msg += f"\t\tExit Code:: {status.state.terminated.exit_code}\n"
+                                    if status.state.terminated.message:
+                                        msg += f"\t\tMessage: {status.state.terminated.message}\n"
+                                    if status.state.terminated.reason:
+                                        msg += f"\t\tReason: {status.state.terminated.reason}\n"
+                                    if status.state.terminated.signal:
+                                        msg += f"\t\tSignal: {status.state.terminated.signal}\n"
+
+                                pod_failure_message += msg
+
+                            # Send pod failure information to flow run logs
+                            self.client.write_run_logs(
+                                [
+                                    dict(
+                                        flow_run_id=flow_run_id,
+                                        name=self.name,
+                                        message=pod_failure_message.strip(),
+                                        level="ERROR",
+                                    )
+                                ]
+                            )
+
+                        if failed_pods:
+                            self.logger.debug(
+                                f"Failing flow run {flow_run_id} due to the failed pods {failed_pods}"
+                            )
+                            self.client.set_flow_run_state(
```

Comment:
I could see cases where the flow run already completed successfully (per the prefect backend) but something happened between reporting that to the backend and the container exiting. I'm not sure if that should trigger a flow run to be marked as failed.

Likewise, I think there are cases where a flow run might have already reported itself as failed (for a flow-level reason) before having it exit with a non-zero exit code. I'm not sure if we want to update the state to failed again.

Perhaps we only want to set the flow run state if it hasn't moved to a finished state already? Not sure if there's an easy way to do this, but it seems like a better user experience to me.